python/shotgun_model/data_handler_find.py

```diff
@@ -0,0 +1,384 @@
+                            })
+                            num_modifications += 1
+
+                else:
+                    # not on leaf level yet
+                    if unique_field_value not in sub_tree[self.CACHE_CHILDREN]:
+                        # item is not yet inserted in our new tree so add it
+                        # because these are parent items like project nodes
+
+                        item = {
+                            self.SG_DATA: sg_item,
+                            self.FIELD: field_name,
+                            self.IS_LEAF: False,
+                            self.UID: unique_field_value,
+                            self.PARENT: sub_tree,
+                            self.CACHE_CHILDREN: {},
+                        }
+                        sub_tree[self.CACHE_CHILDREN][unique_field_value] = item
+                        new_cache[self.CACHE_BY_UID][unique_field_value] = item
+
+                        # now check with prev data structure to see if it has changed
+                        if unique_field_value not in self._cache[self.CACHE_BY_UID]:
+                            # this is a new node that wasn't there before
+                            diff_list.append({
+                                "data": ShotgunDataItem(item),
+                                "mode": self.ADDED
+                            })
+                            num_adds += 1
+                        else:
+                            # record already existed in prev dataset. Check if value has changed
+                            current_record = self._cache[self.CACHE_BY_UID][unique_field_value][self.SG_DATA]
+                            # don't compare the whole record but just the part that relates to this
+                            # intermediate node value. For example, we may be looking at a project node
+                            # in the hierarchy but the full sg record contains all the data for a shot.
+                            # in this case, just run the comparison on the project subset of the full
+                            # shot data dict.
+                            if not self._sg_compare_data(current_record.get(field_name), sg_item.get(field_name)):
+                                diff_list.append({
+                                    "data": ShotgunDataItem(item),
+                                    "mode": self.UPDATED
+                                })
+                                num_modifications += 1
+
+                    # recurse down to the next level
+                    sub_tree = sub_tree[self.CACHE_CHILDREN][unique_field_value]
+
+        # now figure out if anything has been removed
+        self._log_debug("Diffing new tree against old tree...")
+
+        current_uids = set(self._cache[self.CACHE_BY_UID].keys())
+        new_uids = set(new_cache[self.CACHE_BY_UID].keys())
+
+        for deleted_uid in current_uids.difference(new_uids):
```

Comment:
not sure - the data sets are usually very similar in size, i don't think it makes a practical difference. Memory wise i think it's the same.